lib/msprime.c

```diff
@@ -4746,32 +4754,45 @@ msp_run_sweep(msp_t *self)
             rec_rates[j] = recomb_mass;
         }
 
-        curr_step++;
         event_prob = 1.0;
         event_rand = gsl_rng_uniform(self->rng);
         sweep_over = false;
         while (event_prob > event_rand && curr_step < num_steps && !sweep_over) {
             sweep_dt = time[curr_step] - time[curr_step - 1];
             /* using pop sizes grabbed from get_population_size */
             pop_size = get_population_size(&self->populations[0], time[curr_step]);
-            p_coal_B = ((sweep_pop_sizes[1] * (sweep_pop_sizes[1] - 1)))
-                       / allele_frequency[curr_step] * sweep_dt / pop_size;
-            p_coal_b = ((sweep_pop_sizes[0] * (sweep_pop_sizes[0] - 1)))
-                       / (1.0 - allele_frequency[curr_step]) * sweep_dt / pop_size;
+
+            // to avoid # lineages < 2
+            if (avl_count(&self->populations[0].ancestors[1]) > 1)
+                p_coal_B = ((sweep_pop_sizes[1] * (sweep_pop_sizes[1] - 1)))
+                           / allele_frequency[curr_step] * sweep_dt / pop_size;
+            else
+                p_coal_B = 0;
+            // to avoid # lineages < 2
+            if (avl_count(&self->populations[0].ancestors[0]) > 1)
+                p_coal_b = ((sweep_pop_sizes[0] * (sweep_pop_sizes[0] - 1)))
+                           / (1.0 - allele_frequency[curr_step]) * sweep_dt / pop_size;
+            else
+                p_coal_b = 0;
             p_rec_b = rec_rates[0] * sweep_dt;
             p_rec_B = rec_rates[1] * sweep_dt;
             sweep_pop_tot_rate = p_coal_b + p_coal_B + p_rec_b + p_rec_B;
             /* doing this to build in generality if we want >1 pop */
+
             total_rate = sweep_pop_tot_rate;
             event_prob *= 1.0 - total_rate;
             curr_step++;
-            sweep_over = total_rate == 0;
+
+            // Not using total_rate == 0; otherwise crashes happen due to precision
+            // issue.
+            sweep_over = total_rate <= DBL_EPSILON;
```

Comment:
Does this really happen? I thought we'd end up with total_rate = 0 because we multiply by zero at some point. How would we end up with very tiny total_rate values?

Reply:
You are right. This must have happened when I was messing around to locate the bug. It seems changing it back to 
`sweep_over = total_rate == 0;` does not cause any issue in multiple tests.  I will change this line back to the previous version.